covid19_sfbayarea/data/san_mateo/power_bi_querier.py

```diff
@@ -0,0 +1,140 @@
+        self.name = getattr(self, 'name')
+        self.powerbi_resource_key = getattr(self, 'powerbi_resource_key', self.DEFAULT_POWERBI_RESOURCE_KEY)
+        self.property = getattr(self, 'property')
+        self.source = getattr(self, 'source')
+
+    def _fetch_data(self) -> Dict:
+        response = post( self.BASE_URI, headers = { 'X-PowerBI-ResourceKey': self.powerbi_resource_key }, json = self._query_params())
+        response.raise_for_status()
+        return response.json()
+
+    def _parse_data(self, response_json: Dict[str, List]) -> List:
+        results = self._dig_results(response_json)
+        return self._extract_lists(results)
+
+    def _query_params(self) -> Dict[str, Any]:
+        return {
+            'version': '1.0.0',
+            'queries': [self._query()],
+            'cancelQueries': [],
+            'modelId': self.model_id
+        }
+
+    def _query(self) -> Dict[str, Any]:
+        return {
+            'Query': { 'Commands': [self._command()] },
+            'CacheKey': json.dumps({ 'Commands': [self._command()] }),
+            'QueryId': '',
+            'ApplicationContext': {
+                'DatasetId': self.DATASET_ID,
+                'Sources': [{ 'ReportId': self.REPORT_ID }]
+            }
+        }
+
+    def _command(self) -> Dict[str, Dict[str, Dict[str, Any]]]:
+        return {
+             'SemanticQueryDataShapeCommand': {
+                 'Query': {
+                     'Version': 2,
+                     'From': [{ 'Name': self.source, 'Entity': self.name }],
+                     'Select': self._select(),
+                     'OrderBy': self._order_by()
+                 },
+                 'Binding': self._binding()
+             }
+         }
+
+    def _select(self) -> List[Dict[str, Any]]:
+        return [
+            {
+                'Column': self._column_expression(self.property),
+                'Name': f'{self.name}.{self.property}'
+            },
+            {
+                'Aggregation': {
+                    'Expression': { 'Column': self._column_expression('n') },
+                    'Function': 0
+                },
+                'Name': f'CountNonNull({self.name}.n)'
+            }
+       ]
+
+    def _order_by(self) -> List[Dict[str, Any]]:
+        return [
+            {
+                'Direction': 1,
+                'Expression': { 'Column': self._column_expression(self.property) }
+            }
+        ]
+
+    def _column_expression(self, property: str) -> Dict[str, Any]:
+        return {
+            'Expression': { 'SourceRef': { 'Source': self.source } },
+            'Property': property
+        }
+
+    def _binding(self) -> Dict[str, Any]:
+        return {
+            'Primary': { 'Groupings': [{ 'Projections': [0, 1] }] },
+            'DataReduction': {
+                'DataVolume': 4,
+                'Primary': { 'Window': { 'Count': 1000 } }
+            },
+            'Version': 1
+        }
```

Comment:
Is this scraper resilient to changes in the dashboard like 1ec5/external-data-bot@140142ede6ea76ae4c2c597f03a9a0d19822bcc1? I’ve had to update https://github.com/1ec5/external-data-bot/commits/master/scripts/covid19/US/CA/san-mateo-county.sh approximately weekly because of parameters that blit in and out in the dashboard. Any mismatch between parameters causes the `CacheKey` to mismatch, which causes the request to return nothing.

Reply:
Hey, this is a good point. For the time series, cases, and death data, this has been working since I started on it and still works today.

For the testing data, results seem to be limited at May 29th. I'm not sure if this has to do with a cache key or with limiting.

How are you generating your cache keys? The reason for implementing it this way is because I only have to change the query, then the cache key will be regenerated with the `json.dumps` call in the `_query` method.

Reply:
I’ve been manually looking at the XHRs the dashboard uses and just using those requests (stripping out any queries I didn’t need). The requests come with cache keys that I’m using verbatim. Each time the dashboard changes, I’ve been repeating the process. So you’re able to send arbitrary queries to this API that don’t necessarily match the dashboard’s? That would be a game-changer.

Reply:
Exactly. There's actually a bit more to the query that I can remove.

I like the idea of using the actual XHR from the dashboard though. If we could automatically infer what the XHR will be, we could use that without having to do any of this query mangling. I'll check into that. I know Selenium doesn't have an API for it, but there might be another tool or a private API that allows this.